contracts/dynamic-contracts/ExtensionRegistry.sol

```diff
@@ -34,9 +36,31 @@ contract ExtensionRegistry is IExtensionRegistry, ExtensionState, PermissionsEnu
         _updateExtension(_extension);
     }
 
-    /// @notice Remove an existing extension from the registry.
-    function removeExtension(string memory _extensionName) external onlyRole(DEFAULT_ADMIN_ROLE) {
-        _removeExtension(_extensionName);
+    /// @notice Adds an extension to an extension snapshot.
+    function buildExtensionSnapshot(
+        string memory _extensionSnapshotId,
+        string[] memory _extensionNames,
+        bool _freeze
```

Comment:
API is a little bit weird with the fact that i can call it multiple times and then freeze, feels error prone. like, what happens if i call this function again after i freeze? can I unfreeze later? not clear at all

Was there a reason not to a single function that always overrides the passed in set? seems simpler.

with your API, how do I remove an extension from a snapshot?

Reply:
The reason to not make it callable just once is that you may run out of gas when adding `n` number of extensions to a snapshot. I think we want to be able to keep that `n` arbitrary. That's why I opted to let the admin build a snapshot over multiple transactions, and then freeze the snapshot in the final call.

It is important to freeze a snapshot i.e. a 'default set of extensions', since pushing updates to a in-use snapshot would mean we're pushing updates to end-user contracts, as opposed to opt-in upgrades.

Reply:
allowing n calls to build up the snapshot makes sense. I see why you did it this way. Still wondering if we should make freeze a separate call tho. feels weird to call this again with an empty array and freeze = true.

What does the flow look like to update a default extension set for say marketplace? that would require a new snapshot id right?. This means that we need to always encode the current version inside the snapshotId?

Reply:
We should map out what it means to ship an update to an extension with this flow. Say an extension that's used by 3 different snapshots for example, like claim conditions.

Reply:
You wouldn't make a separate freeze call with an empty array. You'll specify `freeze=true` in the final tx i.e. where you're adding your final bunch of extensions to a snapshot.